pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/UnnecessaryCaseChangeRule.java

```diff
@@ -4,83 +4,111 @@
 
 package net.sourceforge.pmd.lang.java.rule.errorprone;
 
-import net.sourceforge.pmd.lang.ast.Node;
+import static java.util.Arrays.asList;
+
+import java.util.List;
+
+import net.sourceforge.pmd.lang.java.ast.ASTArgumentList;
+import net.sourceforge.pmd.lang.java.ast.ASTArguments;
 import net.sourceforge.pmd.lang.java.ast.ASTName;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimarySuffix;
+import net.sourceforge.pmd.lang.java.ast.JavaNode;
 import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
 
 public class UnnecessaryCaseChangeRule extends AbstractJavaRule {
 
+    private static final List<String> CASE_CHANGING_METHODS = asList("toLowerCase", "toUpperCase");
+    private static final List<String> EQUALITY_METHODS = asList("equals", "equalsIgnoreCase");
+
     @Override
-    public Object visit(ASTPrimaryExpression exp, Object data) {
-        int n = exp.getNumChildren();
-        if (n < 4) {
-            return data;
+    public Object visit(ASTPrimaryExpression expr, Object data) {
+        int caseChangingCallIndex = getCaseChangingMethodCallIndex(expr);
+        if (caseChangingCallIndex != -1) {
+            int chainedMethodCallIndex = caseChangingCallIndex + 2;
+            if (hasEqualsMethodCallChainedAtPosition(expr, chainedMethodCallIndex)
+                    || isArgumentOfEqualsMethodCall(expr)) {
+                addViolation(data, expr);
+            }
         }
+        return super.visit(expr, data);
```

Comment:
We seem to visit only one type in this rule, so you can make use of the rulechain: https://pmd.github.io/latest/pmd_userdocs_extending_writing_java_rules.html#economic-traversal-the-rulechain